testcase2owl/phyloref/TaxonomicUnit.py

```diff
@@ -327,9 +321,85 @@ def load_from_jsonld(self, jsonld):
     def from_jsonld(jsonld):
         """ Read a Specimen from a JSON-LD object and return it. """
         sp = Specimen()
-        sp.load_from_json(jsonld)
+        sp.load_from_jsonld(jsonld)
         return sp
 
-    def as_json(self):
+    def as_jsonld(self):
         """ Return this Specimen as a JSON-LD object. """
         return self.properties
+
+    # Helper properties to access some common ways of identifying specimens.
+    @property
+    def identifier(self):
+        """
+        Retrieves a single identifier for this specimen.
+
+        Based on http://dx.doi.org/10.1371/journal.pone.0114069, there appear to be two common schemes
+        for doing this:
+         - institutionCode + ':' + (collectionCode + ':')? + catalogNumber
+         - occurrenceID, preferably expressed as 'urn:catalog:[institutionCode]:[collectionCode]:[catalogNumber]'
+
+        Oddly, they don't refer to materialSampleID; Darwin Core refers to this as a reference to the sample and not
+        just its digital record, so that might be why.
+
+        At the level of the Taxonomic Unit, CDAO provides a third option: setting an external reference to a URI that
+        identifies a specimen, such as:
+         - http://portal.vertnet.org/o/ku/kuh?id=be1b5c81-b069-11e3-8cfe-90b11c41863e
+         - https://www.idigbio.org/portal/records/07e2d181-cc8b-4b6d-9e27-58756ea08c9b
+
+        :return: A combined identifier if one is detectable, or None.
+        """
+
+        # Option 1. Compare occurrenceIDs
+        if 'occurrenceID' in self.properties and self.properties['occurrenceID'].strip() != '':
+            return self.properties['occurrenceID']
+
+        if 'catalogNumber' in self.properties:
+            catalogNumber = self.properties['catalogNumber']
+            collectionCode = self.properties.get('collectionCode')
+            institutionCode = self.properties.get('institutionCode')
+
+            if institutionCode is None:
+                if collectionCode is None:
+                    return catalogNumber
+                else:
+                    return collectionCode + ':' + catalogNumber
+            else:
+                if collectionCode is None:
+                    return institutionCode + ':' + catalogNumber
+                else:
+                    return institutionCode + ':' + collectionCode + ':' + catalogNumber
+
+        return None
+
+    @identifier.setter
+    def identifier(self, value):
+        """ Sets the occurrenceId and look for Darwin Core Triplet components. """
+        self.properties['occurrenceID'] = value
+
+        # Reset the Darwin Core Triplet components.
+        del(self.properties['catalogNumber'])
+        del(self.properties['collectionCode'])
+        del(self.properties['institutionCode'])
+
+        components = value.strip().split(':')
+        if len(components) <= 1:
+            # Nothing we can do.
+            pass
```

Comment:
Well, the getter variant returns a catalog number as a valid value (without colons). Any reason not to symmetrically treat len() == 1 as catalog number?

I guess the problem is distinguishing this from the occurrenceID being a HTTP URI. You can test protocol though.